finality-aleph/src/sync/handler/request_handler.rs

```diff
@@ -0,0 +1,423 @@
+                    match j {
+                        Some(j) => HeadOfChunk::Justification(j, b, None),
+                        None => {
+                            pre_chunk.add_block(b);
+                            HeadOfChunk::Header(self.parent_header(&header)?)
+                        }
+                    }
+                }
+                State::Everything => {
+                    let (j, b, h) = self.get_justification_block_and_header(header.id())?;
+                    match j {
+                        Some(j) => HeadOfChunk::Justification(j, b, h),
+                        None => {
+                            pre_chunk.add_block(b);
+                            pre_chunk.add_header(h);
+                            HeadOfChunk::Header(self.parent_header(&header)?)
+                        }
+                    }
+                }
+                State::OnlyJustification => {
+                    let j = self.get_justification(header.id())?;
+                    match j {
+                        Some(j) => HeadOfChunk::Justification(j, None, None),
+                        None => HeadOfChunk::Header(self.parent_header(&header)?),
+                    }
+                }
+                State::Finished => return Err(RequestHandlerError::BadState),
+            };
+
+            state = match branch_knowledge {
+                BranchKnowledge::LowestId(id) if *id == header.id() => State::Everything,
+                _ => state,
+            };
+
+            if let HeadOfChunk::Justification(_, _, _) = &head_chunk {
+                break;
+            }
+        }
+
+        Ok((state, head_chunk, pre_chunk.into_chunks()))
+    }
+
+    fn chunks(
+        self,
+        head: HeadOfChunk<B, J>,
+        branch_knowledge: BranchKnowledge<J>,
+        to: BlockIdFor<J>,
+    ) -> HandlerResult<Vec<Chunk<B, J>>, CS::Error> {
+        let mut chunks = vec![];
+
+        let mut head = head;
+        let mut state = State::EverythingButHeader;
+
+        while state != State::Finished {
+            let (new_state, new_head, chunks_from_step) =
+                self.step(state, head, &to, &branch_knowledge)?;
+
+            chunks.extend(chunks_from_step);
+            state = new_state;
+            head = new_head;
+        }
+
+        Ok(chunks)
+    }
+
+    pub fn response(self, request: Request<J>) -> HandlerResult<Response<B, J>, CS::Error> {
+        if !request.is_valid() {
+            return Err(RequestHandlerError::BadRequest);
+        }
+
+        let our_top_justification = self.chain_status.top_finalized()?;
+        let top_justification = request.state().top_justification();
+        let target = request.target_id();
+
+        let upper_limit = self.upper_limit(top_justification.id());
+
+        // request too far into future
+        if target.number() > upper_limit {
+            return Ok(Response::from_chunks(vec![]));
```

Comment:
Uh, we probably shouldn't be sending empty responses over the network for absolutely no reason, now should we?

And btw, this should probably be doing something else anyway – you constructed the upper limit from their top justification, so if the target is too far then they didn't really want to request it. Either return a `BadRequest` error or see whether you can respond with something useful by just trying to respond to `upper_limit`.

Reply:
now we return Action for 3 possibilities.